src/scene/layers.js

```diff
@@ -0,0 +1,786 @@
+                this._staticLightHash = 0;
+            } else {
+                this._staticLightHash = pc.hashCode(str);
+            }
+        } else {
+            this._staticLightHash = 0;
+        }
+    };
+
+    Layer.prototype._generateCameraHash = function () {
+        // generate hash to check if cameras in layers are identical
+        // order of cameras shouldn't matter
+        if (this.cameras.length > 1) {
+            this.cameras.sort(sortCameras);
+            var str = "";
+            for(var i=0; i<this.cameras.length; i++) {
+                str += this.cameras[i].entity._guid;
+            }
+            this._cameraHash = pc.hashCode(str);
+        } else {
+            this._cameraHash = 0;
+        }
+        this._dirtyCameras = true;
+    };
+
+    Layer.prototype.addCamera = function (camera) {
+        if (this.cameras.indexOf(camera) >= 0) return;
+        this.cameras.push(camera);
+        this._generateCameraHash();
+    };
+
+    Layer.prototype.removeCamera = function (camera) {
+        var id = this.cameras.indexOf(camera);
+        if (id < 0) return;
+        this.cameras.splice(id, 1);
+        this._generateCameraHash();
+    };
+
+    Layer.prototype.sortCameras = function () {
+        this._generateCameraHash();
+    };
+
+    Layer.prototype._calculateSortDistances = function(drawCalls, drawCallsCount, camPos, camFwd) {
+        var i, drawCall, btype, meshPos;
+        var tempx, tempy, tempz;
+        for (i = 0; i < drawCallsCount; i++) {
+            drawCall = drawCalls[i];
+            if (drawCall.command) continue;
+            if (drawCall.layer <= pc.scene.LAYER_FX) continue; // Only alpha sort mesh instances in the main world (backwards comp)
+            meshPos = drawCall.aabb.center.data;
+            tempx = meshPos[0] - camPos[0];
+            tempy = meshPos[1] - camPos[1];
+            tempz = meshPos[2] - camPos[2];
+            drawCall.zdist = tempx*camFwd[0] + tempy*camFwd[1] + tempz*camFwd[2];
+        }
+    };
+
+    Layer.prototype._sortVisible = function (transparent, cameraNode, cameraPass) {
+        var objects = this.instances;
+        var sortMode = transparent ? this.transparentSortMode : this.opaqueSortMode;
+        if (sortMode === pc.SORTMODE_NONE) return;
+        var visible = transparent ? objects.visibleTransparent[cameraPass] : objects.visibleOpaque[cameraPass];
+        if (sortMode === pc.SORTMODE_BACK2FRONT || sortMode === pc.SORTMODE_FRONT2BACK) {
+            sortPos = cameraNode.getPosition().data;
+            sortDir = cameraNode.forward.data;
+            this._calculateSortDistances(visible.list, visible.length, sortPos, sortDir);
+        }
+
+        if (visible.list.length !== visible.length) {
+            visible.list.length = visible.length;
+        }
+        visible.list.sort(sortCallbacks[sortMode]);
+    };
+
+    // Composition can hold only 2 sublayers of each layer
+    var LayerComposition = function () {
```

Comment:
Can we split this file into to: layer.js and layer-composition.js

Reply:
Done.